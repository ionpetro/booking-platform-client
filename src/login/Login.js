import Button from '../components/Button/Button.vue';
import Input from '../components/Input/Input.vue';
import Message from '../components/Message/Message.vue';
import User from '../models/user';
import AuthService from '../shared/services/auth.service';

const authService = new AuthService();

export default {
  name: 'Login',
  components: {
    'v-button': Button,
    'v-input': Input,
    'v-message': Message
  },
  data() {
    return {
      show: false,
      loading: false,
      user: new User('', ''),
      remember: false,
      errors: {},
      message: ''
    };
  },
  mounted() {
    this.show = true;
  },
  methods: {
    checkForm() {
      this.errors = {};

      if (!this.user.username) {
        this.errors.username = '* Please, fill your username.';
      } else if (!this.validUsername(this.user.username)) {
        this.errors.username = '* Please, correct your username';
      }

      if (!this.user.password) {
        this.errors.password = '* Please, fill your password.';
      }

      // if there are no validation errors, proceed with login
      if (!Object.keys(this.errors).length) {
        this.handleLogin();
      }
    },
    validUsername(username) {
      const re = /^\w{2,8}@(\d{4}\.)[\w]{2}$/;
      return re.test(username);
    },
    handleLogin() {
      this.loading = true;
      if (this.user.username && this.user.password) {
        authService.login(this.user, this.remember).then(
          () => {
            this.$router.push('/');
          },
          error => {
            this.loading = false;
            this.message =
              (error.response && error.response.data) || error.message || error.toString();
          }
        );
      }
    },
    deleteErrors(labelName) {
      // when user focuses on the input, the error message is deleted
      if (labelName === 'Username') {
        this.$delete(this.errors, 'username');
      }
      if (labelName === 'Password') {
        this.$delete(this.errors, 'password');
      }
    }
  },
  created() {
    // Redirect user to home page if already authenticated
    let user = localStorage.getItem('user') || sessionStorage.getItem('user');
    user = JSON.parse(user);
    if (user) {
      this.$router.push('/');
    }
  }
};
