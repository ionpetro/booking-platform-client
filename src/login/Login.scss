.login {
  display: flex;
}

.cancel {
  cursor: pointer;
  position: absolute;
  top: 25px;
  right: 25px;
  transition: transform 0.2s ease-in;
  filter: invert(0.5);

  &:hover {
    transform: scale(1.1);
  }
}

.hero {
  float: left;
  height: 90vh;
  max-height: 1200px;

  // tablet
  @media #{$tablet} {
    display: none;
  }
  // portrait screen
  @media only screen and (min-width: $medium-breakpoint) and (min-height: 1300px) {
    display: none;
  }
  // big screen
  @media only screen and (min-width: 2000px) and (min-height: 1300px) {
    display: none;
  }
  // portrait on big screen
  @media only screen and (min-width: 2000px) and (min-height: 1800px) {
    display: none;
  }
}
.hero-big {
  display: none;
  width: auto;

  // portrait screen
  @media only screen and (min-width: $medium-breakpoint) and (min-height: 1300px) {
    display: none;
  }
  // big screen
  @media only screen and (min-width: 2000px) and (min-height: 1300px) {
    display: block;
  }
  // portrait on big screen
  @media only screen and (min-width: 2000px) and (min-height: 1800px) {
    display: none;
  }
}
.logo {
  display: block;
  margin: 81px auto 0 auto;
}
.heading {
  margin-bottom: 0 !important;
  margin-top: 127px;
  text-align: center;

  // mobile
  @media #{$mobile} {
    margin-top: 48px;
  }
}
.content {
  width: 480px;
  margin: auto;

  // tablet
  @media #{$tablet} {
    width: 480px;
    margin: auto;
  }
  // mobile
  @media #{$mobile} {
    width: 100%;
    margin-left: 16px;
    margin-right: 16px;
  }
  // portrait screen
  @media only screen and (min-width: $medium-breakpoint) and (min-height: 1300px) {
    position: relative;
    top: 30vh;
  }
  // big screen
  @media only screen and (min-width: 2000px) and (min-height: 1300px) {
    position: relative;
    top: -100px;
  }
  // portrait on big screen
  @media only screen and (min-width: 2000px) and (min-height: 1800px) {
    position: relative;
    top: 30vh;
  }
}
.errors {
  color: $red-medium;
  position: absolute;
  padding-left: 15px;
}

.checkbox {
  margin-top: 47px;
  display: flex;
  max-width: 200px;
  color: $secondary;

  &__control {
    display: inline-grid;
    background-color: $white;
    width: 1em;
    height: 1em;
    border-radius: 0.2em;
    border: 0.09rem solid;
    // center it
    margin-top: 0.2rem;

    svg {
      transition: transform 0.1s ease-in 25ms;
      transform: scale(0);
      transform-origin: bottom left;
      margin: auto;
    }
  }

  &__input {
    display: grid;
    grid-template-areas: 'checkbox';

    > * {
      grid-area: checkbox;
    }

    input {
      opacity: 0;
      width: 1em;
      height: 1em;

      &:checked + .checkbox__control svg {
        transform: scale(1);
      }
      &:checked + .checkbox__control {
        background-color: $secondary;
      }
      &:focus-visible + .checkbox__control {
        box-shadow: 0 0 0 2px $orange-light;
      }
    }
  }
}

.button--margin {
  margin-top: 53px;
  margin-left: auto;
  margin-right: auto;
}

// transitions
.fade-enter-active {
  transition: opacity 1s;
}
.fade-enter {
  opacity: 0;
}
