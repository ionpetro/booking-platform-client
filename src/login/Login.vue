<template>
  <div>
    <h1>Login Page</h1>
    <form name="form" @submit.prevent="handleLogin">
      <div>
        <label for="email">Colonist ID</label>
        <input type="text" v-model="user.username" />
      </div>
      <div>
        <label for="password">Password</label>
        <input type="password" v-model="user.password" />
      </div>
      <div>
        <input type="checkbox" />
        <label for="rememberMe">Remember Me</label>
      </div>
      <button type="submit">Sign In</button>
    </form>
  </div>
</template>

<script>
import User from '../models/user';
import AuthService from './auth.service';

const authService = new AuthService();

export default {
  name: 'Login',
  data() {
    return {
      loading: false,
      user: new User('', ''),
      remember: false
    };
  },
  methods: {
    handleLogin() {
      console.log(authService);
      this.loading = true;
      if (this.user.username && this.user.password) {
        authService.login(this.user, this.remember).then(response => {
          this.loading = false;
          return Promise.resolve(response);
        });
      }
    }
  }
};
</script>

<style></style>
