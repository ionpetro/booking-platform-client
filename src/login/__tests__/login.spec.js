import {shallowMount} from '@vue/test-utils';
import flushPromises from 'flush-promises';
import Login from '../Login.vue';
// import authService from '../src/shared/services/auth.service';

// jest.mock('../src/shared/services/auth.service');

describe('Login', () => {
  let wrapper;
  const $router = {push: jest.fn()};

  beforeEach(() => {
    wrapper = shallowMount(Login, {
      mocks: {$router}
    });
  });

  const findBtnSignIn = () => wrapper.find('button');
  const findInputUsername = () => wrapper.find('#username > input');
  const findInputPassword = () => wrapper.find('#password > input');
  const findMsgError = () => wrapper.find('.error');
  const findForm = () => wrapper.find('form');

  it('has data', () => {
    expect(typeof Login.data).toBe('function');
  });

  describe('when loaded', () => {
    it('has the required elements', () => {
      const findTitle = () => wrapper.find('h1');

      expect(findTitle().exists()).toBe(true);
      expect(findTitle().text()).toBe('Login Page');
      expect(findInputUsername().exists()).toBe(true);
      expect(findForm().exists()).toBe(true);
      expect(findInputPassword().exists()).toBe(true);
      expect(findBtnSignIn().exists()).toBe(true);
      expect(findBtnSignIn().text()).toBe('Sign In');
      expect(findMsgError().exists()).toBe(false);
    });
  });

  describe('when sign in button is clicked', () => {
    // helper function
    const fillLoginFieldAndSubmit = async (username, password) => {
      findInputUsername().setValue(username);
      findInputPassword().setValue(password);
      findForm().trigger('submit.prevent');
      await flushPromises();
    };
    const assertErrorMessage = message => {
      expect(findMsgError().exists()).toBe(true);
      expect(findMsgError().text()).toBe(message);
    };

    it('shows error when username or password is empty', async () => {
      await fillLoginFieldAndSubmit('', 'test');
      assertErrorMessage('Username Required');

      await fillLoginFieldAndSubmit('test@1234.co', '');
      assertErrorMessage('Password Required');

      await fillLoginFieldAndSubmit('test', 'test');
      assertErrorMessage('Valid Username Required');
    });

    it('hits login API and redirects to homepage', async () => {
      // TODO: Test this feature
      // authService.login.mockResolvedValue();
      // await fillLoginFieldAndSubmit('test@1234.co', 'test');
      // expect(authService.login).toHaveBeenCalled();
      // expect($router.push).toBeCalledWith('home');
    });

    it('shows error when API hit throws error', async () => {
      // TODO: Test this feature
      //   authService.login.mockRejectedValue();
      //   await fillLoginFieldAndSubmit('test@1234.co', 'test');
      //   expect(service.login).toBeCalled();
      //   assertErrorMessage('Login failed');
    });

    it('handlesLogin when both username and password are valid', async () => {
      await wrapper.find('#username > input').setValue('joe@1234.co');
      await wrapper.find('#password > input').setValue('bestPassw0rd');

      await wrapper.find('form').trigger('submit.prevent');
      expect(wrapper.vm.loading).toBe(true);
    });
  });
});
