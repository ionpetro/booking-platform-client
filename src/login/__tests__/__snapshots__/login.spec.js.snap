// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login when loaded renders correctly 1`] = `
<transition-stub
  name="fade"
>
  <div
    class="login"
  >
    <div
      class="cancel"
    >
      <img
        alt="Cancel icon"
        src="@/assets/icons/Remove.svg"
      />
    </div>
     
    <div
      class="hero-container"
    >
      <img
        alt="hero"
        class="hero"
        src="@/assets/images/hero.png"
      />
       
      <img
        alt="hero-big"
        class="hero-big"
        src="@/assets/images/hero-big.png"
      />
    </div>
     
    <div
      class="content"
    >
      <img
        alt="logo"
        class="logo"
        src="@/assets/images/blue.png"
      />
       
      <h1
        class="heading"
      >
        Welcome to Mars era
      </h1>
       
      <transition-stub
        name="fade"
      >
        <!---->
      </transition-stub>
       
      <form
        name="form"
      >
        <div
          id="username"
        >
          <div
            class="input username"
          >
            <input
              class="input__text"
              placeholder="ex. joe@2020.co"
              type="text"
            />
             
            <label
              class="label__text label__text"
              for="name"
            >
              Username
            </label>
             
            <!---->
          </div>
           
          <!---->
        </div>
         
        <div
          id="password"
        >
          <div
            class="input password"
          >
            <input
              class="input__text"
              placeholder="Fill your password here.."
              type="password"
            />
             
            <label
              class="label__text label__text"
              for="name"
            >
              Password
            </label>
             
            <div
              class="input__icon"
              tabindex="0"
            >
              <img
                alt="show password"
                src="@/assets/icons/Eye.svg"
              />
               
              <img
                alt="hide password"
                src="@/assets/icons/Eye-Off.svg"
                style="display: none;"
              />
            </div>
          </div>
           
          <!---->
        </div>
         
        <div
          id="remember"
        >
          <label
            class="checkbox"
          >
            <span
              class="checkbox__input"
            >
              <input
                name="checkbox"
                type="checkbox"
              />
               
              <span
                class="checkbox__control"
              >
                <svg
                  fill="none"
                  height="9"
                  viewBox="0 0 9 9"
                  width="9"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8.16583 0.833374L3.24746 7.85928C3.11582 8.04845 2.90126 8.1628 2.67083 8.16661C2.44039 8.17042 2.22218 8.06322 2.08435 7.87851L0.833252 6.21059"
                    stroke="white"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="1.5"
                  />
                </svg>
              </span>
            </span>
             
            <span
              class="checkbox__text"
            >
              Remember me
            </span>
          </label>
        </div>
         
        <button
          class="btn button--margin button primary"
        >
          <span>
            Login
          </span>
           
          <span>
            <!---->
          </span>
        </button>
      </form>
    </div>
  </div>
</transition-stub>
`;
